src/Orleans/Core/GrainFactory.cs

```diff
@@ -30,24 +30,40 @@ MIT License
 
 namespace Orleans
 {
+    using System.Linq;
+    using System.Linq.Expressions;
+    using System.Reflection;
+    using System.Threading;
     using System.Threading.Tasks;
 
+    using GrainInterfaceData = Orleans.CodeGeneration.GrainInterfaceData;
+
     /// <summary>
     /// Factory for accessing grains.
     /// </summary>
     public class GrainFactory : IGrainFactory
     {
         /// <summary>
+        /// The initialization lock.
+        /// </summary>
+        private static readonly object IntitializationLock = new object();
+
+        /// <summary>
+        /// The runtime code generator.
+        /// </summary>
+        private static IRuntimeCodeGenerator codeGenerator;
+
+        /// <summary>
         /// The collection of <see cref="IGrainObserver"/> <c>CreateObjectReference</c> delegates.
         /// </summary>
         private readonly ConcurrentDictionary<Type, Delegate> referenceCreators =
             new ConcurrentDictionary<Type, Delegate>();
 
         /// <summary>
-        /// The collection of <see cref="IGrainObserver"/> <c>DeleteObjectReference</c> delegates.
+        /// The collection of <see cref="IGrainMethodInvoker"/>s for their corresponding grain interface type.
         /// </summary>
-        private readonly ConcurrentDictionary<Type, Delegate> referenceDestoyers =
```

Comment:
From what I can tell `GrainReference.DeleteObjectReference` behaves identically for all `GrainReference` subclasses.